### PR TITLE
added ninja tool for scons

### PR DESCRIPTION
**Pull request by dmoody256**
This is a project localized version from the scons builtin tool I have in my branch (plus some latest cherry picks from mongodb): https://github.com/dmoody256/scons/tree/ninja_scons_daemon

The origin of this tool has been used extensively in the mognodb build by all its developers over the last several years. This is a bleeding edge version featuring a scons daemon which facilitates callback building to scons. This bleeding edge version will make its way in to scons mainline probably within a few months. 

Hopefully godot can start using it and we can flush out any issues so that the mainline version will work out of the box for godot.

Using the tool is simple by just passing `use_ninja=yes` on the scons build command, and a ninja file will be generated.
From there you can just run `ninja` to build. Anytime you change a SConstruct or SCsub or generator py files, it should regenerate ninja file automatically.

Note that the scons daemon is launched by ninja and is running scons in the background waiting for ninja to send build commands that only scons know how to deal with. This would be any kind of  SCons function action or more complicated actions. The daemon uses the SCons environment variable NINJA_SCONS_DAEMON_KEEP_ALIVE to determine how many seconds the scons daemon should stay awake without receiving a command from ninja. The default is 1800 seconds. A new daemon will be used when the ninja file is regenerated.

Also note this requires using scons 4.2.

Also note that things built directly by ninja will not use SCons cache, but the callback nodes are still able to.

I look forward to providing support and bugfix to any issues the godot devs may find when using this tool.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -390,3 +390,8 @@ gcov.css
 # https://clangd.llvm.org/ cache folder
 .clangd/
 .cache/
+
+# ninja tool
+.ninja
+build.ninja
+run_ninja_env.bat
```
Comment (Calinou) [inline .gitignore:397]: Add a missing newline at the end of file (as it was before):

```suggestion
run_ninja_env.bat

```

**Comment by dmoody256** [comment]
Should I add a github action for building with ninja?

**Comment by Calinou** [comment]
> @Calinou Should I add a github action for building with ninja?

CI is pretty busy as it is, so I would prefer *replacing* one of the current CI workflows to use Ninja instead of SCons.

**Comment by dmoody256** [comment]
Im not that well versed with github actions so hopefully I set it up right.

**Comment by dmoody256** [comment]
okay fixed the ninja ci in 
9a61435

**Comment by dmoody256** [comment]
I think also the optimizations for MD5-timestamp and implicit cache should be removed. In my opinion they are not worth it, they save a a very small amount of time but can lead to very painful and time consuming headaches if the user is not fully aware what is going on. With ninja solving the incremental build speed issue, the scons build should be reserved for absolute build correctness.

**Comment by akien-mga** [comment]
> I think also the optimizations for MD5-timestamp and implicit cache should be removed. In my opinion they are not worth it, they save a a very small amount of time but can lead to very painful and time consuming headaches if the user is not fully aware what is going on. With ninja solving the incremental build speed issue, the scons build should be reserved for absolute build correctness.

I would move this to a dedicated PR so that it can be reviewed and merged independently. I do think removing these might be good, they seem to cause more harm than gain.

File: SConstruct [hunk 2]
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env python
 
-EnsureSConsVersion(3, 0, 0)
+EnsureSConsVersion(4, 2, 0)
```
Comment (akien-mga) [inline SConstruct:3]: That's a bit too steep, we still support building Godot on fairly old distros which may not have SCons 4.2.0 packaged yet (as a matter of fact almost no distro has it packaged *right now* - only Arch Linux and Fedora).

There are Godot packages in Linux distro and they rely on system SCons, so they can't just pull it from the pip repos.

Instead, you should check the SCons version at runtime to conditionally support ninja build for SCons 4.2.0 or later, as we do for the compilation DB:

https://github.com/godotengine/godot/blob/3ca2a48767486a52bf4ec5b784a25a1beb2a3f50/SConstruct#L355-L362

(Might also be worth considering if this API or a similar one could be promoted to public in SCons upstream.)

File: SConstruct [hunk 3]
```diff
@@ -1,6 +1,6 @@
...
 EnsurePythonVersion(3, 5)
```
Comment (akien-mga) [inline SConstruct:4]: Not for this PR specifically, but a note to self: SCons 4.2.0 is deprecated Python 3.5 support, so we might want to follow suit and get ready for 4.3.0 by bumping this to 3.6 (and then start porting the codebase to finally benefit from f-strings).

File: site_scons/site_tools/scons_ninja/Globals.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
+# MIT License
+#
+# Copyright The SCons Foundation
```
Comment (akien-mga) [inline site_scons/site_tools/scons_ninja/Globals.py:3]: I'm a bit picky about the folder structure of the Godot repository, and I like the top-level folders to be all Godot code or documentation - buildsystem, CI, and packaging tooling go to `misc` usually.

Could you put this in `misc/scons/scons_ninja/` using the `toolpath` argument in `env.Tool()`? As we used to do in #38830 before the compilation db was upstreamed.

Same goes for other PRs adding tools to SCons' default `site_scons/site_tools/`.

File: SConstruct [hunk 5]
```diff
@@ -123,6 +125,7 @@ opts.Add(EnumVariable("float", "Floating-point precision", "default", ("default"
 opts.Add(EnumVariable("optimize", "Optimization type", "speed", ("speed", "size", "none")))
 opts.Add(BoolVariable("production", "Set defaults to build Godot for use in production", False))
 opts.Add(BoolVariable("use_lto", "Use link-time optimization", False))
+opts.Add(BoolVariable("use_ninja", "Generate a ninja file to build.", False))
```
Comment (akien-mga) [inline SConstruct:128]: To be consistent with other options:
```suggestion
opts.Add(BoolVariable("use_ninja", "Generate a ninja file to build", False))
```

File: SConstruct (hunk 2 above)
Comment (dmoody256) [reply SConstruct:3]: Okay that's reasonable. I'll add a similar conditional check and file something for something more standard upstream.

**Comment by dmoody256** [comment]
Some metrics:

Intel(R) Xeon(R) Platinum 8124M CPU @ 3.00GHz (16 cores)
32 GB ram
Ubuntu 18.04
NVM filesystem

command:
`scons -j16 verbose=true tools=yes tests=yes target=release_debug warnings=extra werrro=yes debug_symbols=no module_text_server_fb_enabled=yes`

ninja:
scons ninja generation: 5.8s
ninja build: 8m9s

pure scons:
8m11s

File: SConstruct (hunk 5 above)
Comment (dmoody256) [reply SConstruct]: done in 
e3b3d25

File: site_scons/site_tools/scons_ninja/Globals.py (hunk 4 above)
Comment (dmoody256) [reply site_scons/site_tools/scons_ninja/Globals.py]: done in 
e3b3d25

File: SConstruct (hunk 3 above)
Comment (dmoody256) [reply SConstruct]: done in 
e3b3d25

File: SConstruct (hunk 2 above)
Comment (dmoody256) [reply SConstruct]: add check like so in
e3b3d25

**Comment by dmoody256** [comment]
> That makes sense. I think we should do like mongodb here and not use ninja in CI, at least in this PR. We can reassess once cache support is improved and if it can be made to play nice with the current GitHub Actions cache.

removed ninja CI in
e3b3d25

**Comment by dmoody256** [comment]
>  Is there a way to improve incremental compilation when the daemon has expired? Systematically rebuilding half the code when there was no change is not usable.

technically the daemon can stay up forever. I made the timeout because the daemon pid file lives in build directory. This is not great, because if you clean you build folder, then the pidfile is lost and the daemon will stay until someone manually finds it and takes it down. It definitely needs some work to make it a better behaved daemon.

The main thing is that if the sconscripts change you need to restart the daemon. For scons specific jobs, ninja will not be able to determine if they need to be rebuilt, so ninja will execute those jobs, and send them to the scons daemon, which will then determine if they really need to be rebuilt. The ninja only jobs shouldn't need to be rebuilt, but maybe there is an issue there. I will take a look. Thanks for taking a closer look!

**Comment by dmoody256** [comment]
@akien-mga So I have a few fixups in my local branch for some of the issues you presented, but still I am getting a case where ninja is rebuilding files it should not after regenerating the ninja file takes place, for example, making a non functional change to SConstruct. I realized there are several parts of the godot build which will re-write files just from scons execution, that is to say these files are generated outside the build graph, during sconscript reading phase. This will not work for the ninja tool, because the ninja tool subverts scons normal build processes to make sure nothing is built during the actual SCons build and only the dependency graph and command lines are generated.

Pure SCons build can be smart about this generated during reading phase because it does MD5 checking and can see that even though the timestamp changed, the MD5 is the same so the file is not actually dirty.

Here is just one example I am talking about: https://github.com/godotengine/godot/blob/6ad0ca8fcd08a5414788fdefc428d92b3f4a67a9/SConstruct#L720-L724

When I was working on a [branch to move all build output to target directory](https://github.com/dmoody256/godot/tree/out_tree_builds), I ended up fixing all the cases I found where things are being generated outside of scons, because in that case too, SCons can not act on things done during SConscript reading phase.

TLDR: I need make some changes to the structure of the build for it work with ninja, namely fixing cases were building is done during sconscript reading phase, should I do that in a separate PR? regardless of ninja, that practice is incorrect form and those build files should become official scons nodes.

**Comment by dmoody256** [comment]
@akien-mga Also I wanted to mention that the ninja tool has merged to upstream scons. That version does not include the scons daemon, and when ninja needs scons to build something, it aggregates the targets, and kicks them all off at once, which has potential to lead to some ordering issues depending on the build structure.

But in any case, while appreciate the help getting the scons daemon feature working, there are other avenues to get ninja working for the godot scons build. Most of godot source code generation takes place in function actions, but it may be possible to outsource these to dedicated python scripts (you would have to do something like this for Meson anyways) and then just have ninja execute the python scripts as a command.
